File: Archive/trumpetdotnet/Controllers/ItemsController.cs
using Microsoft.AspNetCore.Mvc;
using Trumpet.Backend.Models;
using Trumpet.Backend.Services.Items;
using Trumpet.Backend.Services.Tagging;
using Trumpet.Backend.Services.Recommendations;
using Microsoft.AspNetCore.Cors;

namespace Trumpet.Backend.Controllers;

[ApiController]
[EnableCors("AllowAll")]
[Route("api/[controller]")]
public class ItemsController : ControllerBase
{
    private readonly IItemsService _itemsService;
    private readonly ITaggingService _taggingService;
    private readonly IRecommendationService _recommendationService;

    public ItemsController(IItemsService itemsService, ITaggingService taggingService, IRecommendationService recommendationService)
    {
        _itemsService = itemsService;
        _taggingService = taggingService;
        _recommendationService = recommendationService;
    }

    [HttpGet("fields")]
    public async Task<ActionResult<IEnumerable<string>>> GetFields()
    {
        return Ok(await _itemsService.GetFieldsAsync());
    }

    [HttpGet("path-values")]
    public async Task<ActionResult<IEnumerable<string>>> GetPathValues()
    {
        return Ok(await _itemsService.GetPathValuesAsync());
    }

    [HttpGet("path-counts")]
    public async Task<ActionResult<IEnumerable<object>>> GetPathCounts()
    {
        return Ok(await _itemsService.GetPathCountsAsync());
    }

    [HttpGet("search-all")]
    public async Task<ActionResult<IEnumerable<object>>> SearchAllMetadata([FromQuery] string value)
    {
        return Ok(await _itemsService.SearchAllMetadataAsync(value));
    }

    [HttpGet]
    public async Task<ActionResult<IEnumerable<Item>>> GetItems(
        [FromQuery] string? path, 
        [FromQuery] string? search, 
        [FromQuery] string? communityId,
        [FromQuery] string? collectionId,
        [FromQuery] int page = 1, 
        [FromQuery] int pageSize = 10)
    {
        return Ok(await _itemsService.GetItemsAsync(path, search, communityId, collectionId, page, pageSize));
    }

    [HttpGet("{id}")]
    public async Task<ActionResult<Item>> GetItem(string id)
    {
        var item = await _itemsService.GetItemAsync(id);

        if (item == null)
        {
            return NotFound();
        }

        // Ensure tags are loaded/displayed in Metadata (they are part of MetadataValues so they should be there)
        return Ok(item);
    }

    [HttpPost("{id}/tags/generate")]
    public async Task<ActionResult<List<string>>> GenerateTags(string id)
    {
        var item = await _itemsService.GetItemAsync(id);
        if (item == null) return NotFound();

        var tags = await _taggingService.GenerateTagsAsync(item);
        // Note: The service doesn't save them automatically in this method reference, 
        // but for the sake of the API we might want to save them here or have the service do it.
        // The Backfill method saves. The GenerateTagsAsync just returns.
        // Let's implement saving here for single item.
        
        // Actually, let's just use the backfill logic or similar manually here for now, 
        // to avoid complicating the Interface. Ideally TaggingService.AssignTags(sessionId)
        // For MVP: Just return them.
        return Ok(tags);
    }

    [HttpPost("tags/generate-all")]
    public async Task<ActionResult> GenerateAllTags()
    {
        await _taggingService.BackfillTagsAsync();
        return Ok("Tag generation started/completed.");
    }

    [HttpGet("{id}/recommendations")]
    public async Task<ActionResult<IEnumerable<Item>>> GetRecommendations(string id)
    {
        var recommendations = await _recommendationService.GetRecommendationsAsync(id);
        return Ok(recommendations);
    }
}

File: data/backend/src/Controllers/ItemsController.php
<?php

declare(strict_types=1);

namespace Trumpet\Controllers;

use PDO;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Trumpet\Services\ItemsService;
use Trumpet\Services\TaggingService;
use Trumpet\Services\RecommendationService;

class ItemsController
{
    private ItemsService          $itemsService;
    private TaggingService        $taggingService;
    private RecommendationService $recommendationService;

    public function __construct(PDO $pdo, array $settings)
    {
        $this->itemsService          = new ItemsService($pdo);
        $this->taggingService        = new TaggingService($pdo, $settings);
        $this->recommendationService = new RecommendationService($pdo);
    }

    // ── GET /api/items/fields ────────────────────────────────────────────────

    public function getFields(ServerRequestInterface $req, ResponseInterface $res): ResponseInterface
    {
        return $this->json($res, $this->itemsService->getFields());
    }

    // ── GET /api/items/path-values ───────────────────────────────────────────

    public function getPathValues(ServerRequestInterface $req, ResponseInterface $res): ResponseInterface
    {
        return $this->json($res, $this->itemsService->getPathValues());
    }

    // ── GET /api/items/path-counts ───────────────────────────────────────────

    public function getPathCounts(ServerRequestInterface $req, ResponseInterface $res): ResponseInterface
    {
        return $this->json($res, $this->itemsService->getPathCounts());
    }

    // ── GET /api/items/search-all?value=xxx ─────────────────────────────────

    public function searchAllMetadata(ServerRequestInterface $req, ResponseInterface $res): ResponseInterface
    {
        $value = $req->getQueryParams()['value'] ?? '';
        if ($value === '') {
            return $this->json($res, [], 400);
        }
        return $this->json($res, $this->itemsService->searchAllMetadata($value));
    }

    // ── GET /api/items ───────────────────────────────────────────────────────

    public function getItems(ServerRequestInterface $req, ResponseInterface $res): ResponseInterface
    {
        $q = $req->getQueryParams();
        $items = $this->itemsService->getItems(
            path:         $q['path']         ?? null,
            search:       $q['search']       ?? null,
            communityId:  $q['communityId']  ?? null,
            collectionId: $q['collectionId'] ?? null,
            page:         (int) ($q['page']     ?? 1),
            pageSize:     (int) ($q['pageSize'] ?? 10)
        );
        return $this->json($res, $items);
    }

    // ── GET /api/items/{id} ──────────────────────────────────────────────────

    public function getItem(ServerRequestInterface $req, ResponseInterface $res, array $args): ResponseInterface
    {
        $item = $this->itemsService->getItem($args['id']);
        if ($item === null) {
            return $this->json($res, ['detail' => 'Item not found'], 404);
        }
        return $this->json($res, $item);
    }

    // ── POST /api/items/{id}/tags/generate ───────────────────────────────────

    public function generateTags(ServerRequestInterface $req, ResponseInterface $res, array $args): ResponseInterface
    {
        $item = $this->itemsService->getItem($args['id']);
        if ($item === null) {
            return $this->json($res, ['detail' => 'Item not found'], 404);
        }
        $tags = $this->taggingService->generateTags($item);
        return $this->json($res, $tags);
    }

    // ── POST /api/items/tags/generate-all ────────────────────────────────────

    public function generateAllTags(ServerRequestInterface $req, ResponseInterface $res): ResponseInterface
    {
        $count = $this->taggingService->backfillTags();
        return $this->json($res, ['message' => "Tag generation completed for {$count} items."]);
    }

    // ── GET /api/items/{id}/recommendations ──────────────────────────────────

    public function getRecommendations(ServerRequestInterface $req, ResponseInterface $res, array $args): ResponseInterface
    {
        $items = $this->recommendationService->getRecommendations($args['id']);
        return $this->json($res, $items);
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    private function json(ResponseInterface $res, mixed $data, int $status = 200): ResponseInterface
    {
        $res->getBody()->write(json_encode($data, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES));
        return $res
            ->withHeader('Content-Type', 'application/json')
            ->withStatus($status);
    }
}
